apps/arweave/src/ar_verify_chunks.erl

```diff
@@ -0,0 +1,544 @@
+%%% Public interface.
+%%%===================================================================
+
+%% @doc Start the server.
+start_link(Name, StorageModule) ->
+	gen_server:start_link({local, Name}, ?MODULE, StorageModule, []).
+
+-spec name(binary()) -> atom().
+name(StoreID) ->
+	list_to_atom("ar_verify_chunks_" ++ ar_storage_module:label_by_id(StoreID)).
+
+%%%===================================================================
+%%% Generic server callbacks.
+%%%===================================================================
+
+init(StoreID) ->
+	?LOG_INFO([{event, verify_chunk_storage_started}, {store_id, StoreID}]),
+	{StartOffset, EndOffset} = ar_storage_module:get_range(StoreID),
+	gen_server:cast(self(), verify),
+	{ok, #state{
+		store_id = StoreID,
+		packing = ar_storage_module:get_packing(StoreID),
+		start_offset = StartOffset,
+		end_offset = EndOffset,
+		cursor = StartOffset,
+		ready = is_ready(EndOffset),
+		verify_report = #verify_report{
+			start_time = erlang:system_time(millisecond)
+		}
+	}}.
+
+handle_cast(verify, #state{ready = false, end_offset = EndOffset} = State) ->
+	ar_util:cast_after(1000, self(), verify),
+	{noreply, State#state{ready = is_ready(EndOffset)}};
+handle_cast(verify,
+		#state{cursor = Cursor, end_offset = EndOffset} = State) when Cursor >= EndOffset ->
+	ar:console("Done!~n"),
+	{noreply, State};
+handle_cast(verify, State) ->
+	State2 = verify(State),
+	State3 = report_progress(State2),
+	{noreply, State3};
+
+handle_cast(Cast, State) ->
+	?LOG_WARNING([{event, unhandled_cast}, {module, ?MODULE}, {cast, Cast}]),
+	{noreply, State}.
+
+handle_call(Call, From, State) ->
+	?LOG_WARNING([{event, unhandled_call}, {module, ?MODULE}, {call, Call}, {from, From}]),
+	{reply, ok, State}.
+
+handle_info(Info, State) ->
+	?LOG_WARNING([{event, unhandled_info}, {module, ?MODULE}, {info, Info}]),
+	{noreply, State}.
+
+terminate(_Reason, _State) ->
+	ok.
+
+is_ready(EndOffset) ->
+	case ar_block_index:get_last() of
+		'$end_of_table' ->
+			false;
+		{WeaveSize, _Height, _H, _TXRoot}  ->
+			WeaveSize >= EndOffset
+	end.
+
+verify(State) ->
+	#state{store_id = StoreID} = State,
+	{UnionInterval, Intervals} = query_intervals(State),
+	State2 = verify_chunks(UnionInterval, Intervals, State),
+	case State2#state.cursor >= State2#state.end_offset of
+		true ->
+			ar:console("Done verifying ~s!~n", [StoreID]),
+			?LOG_INFO([{event, verify_chunk_storage_verify_chunks_done}, {store_id, StoreID}]);
+		false ->
+			gen_server:cast(self(), verify)
+	end,
+	State2.
+
+verify_chunks(not_found, _Intervals, State) ->
+	State#state{ cursor = State#state.end_offset };
+verify_chunks({End, _Start}, _Intervals, #state{cursor = Cursor} = State) when Cursor >= End ->
+	State;
+verify_chunks({IntervalEnd, IntervalStart}, Intervals, State) ->
+	#state{cursor = Cursor, store_id = StoreID} = State,
+	Cursor2 = max(IntervalStart, Cursor),
+	ChunkData = ar_data_sync:get_chunk_by_byte({chunks_index, StoreID}, Cursor2+1),
```

Comment:
get_chunk_by_byte accepts an offset >= Start, < End so passing Cursor2 should work

Reply:
Hm, it turns out that removing the `+1` caused the script to hang. Not sure why, but I re-added the +1

Reply:
Ah, I was wrong, sorry, it needs > Start, =< End